### ISSUE TITLE
sigsegv in 5.10.1

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10523]
Migrated from [rt.perl.org#76870](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=76870) (status was 'resolved')

Searchable as RT76870$

**Comment by p5pRT** [comment]
### From jettero@cpan.org

#### Created by jettero@cpan.org


This is a bug report for perl from jettero@&#8203;cpan\.org\,
generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
This bug is probably known and fixed\.  I appologize in advance
that I don't really know where to check for that\.  I did flip
through some git commits and perl deltas trying to decide if the
bug was really known or incidentally fixed while acheiving other
goals\.  In any case\, it doesn't apply to more modern perls\, but
5\.10\.x seems to be deployed everywhere so it is perhaps worth
pointing out\.

\#\!/usr/bin/perl

eval \{ \("3"\."x"x25\) =~ m/\(\\d\+\)\(??\{ \_noexist\($^N\) \}\)/; 1 \}

&nbsp;    \# the sagfault will occur in whatever happens to be the next op

&nbsp;    or warn $@&#8203;; \# SIGSEGV in Perl\_pp\_warn\(\) or comment line for
&nbsp;    ;print "here\\n"; \# SIGSEGV in Perl\_do\_print\(\) instead


\# 5\.6\.1  seems unaffected
\# 5\.8\.1  seems to work but complains&#8203;: panic&#8203;: pad\_free curpad\. after "here"
\# 5\.10\.1 where I noticed this \(ubuntu 10\.04\)
\# 5\.12   seems fine
\# 5\.13   also fine

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.1:

Configured by Debian Project at Fri Apr 23 08:18:41 UTC 2010.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=2.6.24-27-server, archname=x86_64-linux-gnu-thread-multi
    uname='linux crested 2.6.24-27-server #1 smp fri mar 12 01:23:09 utc 2010 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.1 -Dsitearch=/usr/local/lib/perl/5.10.1 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.1 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.11.1.so, so=so, useshrplib=true, libperl=libperl.so.5.10.1
    gnulibc_version='2.11.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.10.1:
    /etc/perl
    /usr/local/lib/perl/5.10.1
    /usr/local/share/perl/5.10.1
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.1:
    HOME=/home/jettero
    LANG=en_US.utf8
    LANGUAGE (unset)
    LD_LIBRARY_PATH=
    LOGDIR (unset)
    PATH=/home/jettero/bin:/home/jettero/sbin:/home/jettero/code/povray/bin:/home/jettero/.jbash/jbin:/home/jettero/bin:/home/jettero/sbin:/usr/local/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/games/bin:/usr/X11R6/bin:/home/jettero/code/povray/bin:/usr/local/sbin:/usr/games:/home/jettero/code/apps/bakmud:/home/jettero/code/apps/bdj:/home/jettero/code/apps/convert_videos:/home/jettero/code/apps/ffiec:/home/jettero/code/apps/google.api:/home/jettero/code/apps/jump:/home/jettero/code/apps/mrsh:/home/jettero/code/apps/traffic-view:/home/jettero/code/apps/videodump-pl
    PERL_AUTOINSTALL=--defaultdeps
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From alex@chmrr.net

At Sat Jul 31 08&#8203;:31&#8203;:51 \-0400 2010\, jettero@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> This bug is probably known and fixed\.  I appologize in advance
> that I don't really know where to check for that\.  I did flip
> through some git commits and perl deltas trying to decide if the
> bug was really known or incidentally fixed while acheiving other
> goals\.  In any case\, it doesn't apply to more modern perls\, but
> 5\.10\.x seems to be deployed everywhere so it is perhaps worth
> pointing out\.
> 
> \#\!/usr/bin/perl
> 
> eval \{ \("3"\."x"x25\) =~ m/\(\\d\+\)\(??\{ \_noexist\($^N\) \}\)/; 1 \}
> 
>     \# the sagfault will occur in whatever happens to be the next op
> 
>     or warn $@&#8203;; \# SIGSEGV in Perl\_pp\_warn\(\) or comment line for
>     ;print "here\\n"; \# SIGSEGV in Perl\_do\_print\(\) instead
> 
> 
> \# 5\.6\.1  seems unaffected
> \# 5\.8\.1  seems to work but complains&#8203;: panic&#8203;: pad\_free curpad\. after "here"
> \# 5\.10\.1 where I noticed this \(ubuntu 10\.04\)
> \# 5\.12   seems fine
> \# 5\.13   also fine

This bug seems to be only trigged with \-Dusethreads\, and was fixed by
f7461760003db2ce68155c97ea6c1658e96fcd27\.  Unfortunately\, as that was
a fairly complex feature addition\, it's not a good contender to be
cherry\-picked into maint\-5\.12\.
 \- Alex
\-\- 
Networking \-\- only one letter away from not working

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@chorny - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Sun\, Aug 01\, 2010 at 09&#8203;:19&#8203;:12PM \-0400\, Alex Vandiver wrote&#8203;:
> At Sat Jul 31 08&#8203;:31&#8203;:51 \-0400 2010\, jettero@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> > This bug is probably known and fixed\.  I appologize in advance
> > that I don't really know where to check for that\.  I did flip
> > through some git commits and perl deltas trying to decide if the
> > bug was really known or incidentally fixed while acheiving other
> > goals\.  In any case\, it doesn't apply to more modern perls\, but
> > 5\.10\.x seems to be deployed everywhere so it is perhaps worth
> > pointing out\.
> > 
> > \#\!/usr/bin/perl
> > 
> > eval \{ \("3"\."x"x25\) =~ m/\(\\d\+\)\(??\{ \_noexist\($^N\) \}\)/; 1 \}
> > 
> >     \# the sagfault will occur in whatever happens to be the next op
> > 
> >     or warn $@&#8203;; \# SIGSEGV in Perl\_pp\_warn\(\) or comment line for
> >     ;print "here\\n"; \# SIGSEGV in Perl\_do\_print\(\) instead
> > 
> > 
> > \# 5\.6\.1  seems unaffected
> > \# 5\.8\.1  seems to work but complains&#8203;: panic&#8203;: pad\_free curpad\. after "here"
> > \# 5\.10\.1 where I noticed this \(ubuntu 10\.04\)
> > \# 5\.12   seems fine
> > \# 5\.13   also fine
> 
> This bug seems to be only trigged with \-Dusethreads\, and was fixed by
> f7461760003db2ce68155c97ea6c1658e96fcd27\.  Unfortunately\, as that was
> a fairly complex feature addition\, it's not a good contender to be
> cherry\-picked into maint\-5\.12\.

Valgrind shows that there's still an underlying issue in blead\, it just
happens to no longer manifest itself as a SEGV\. I'm re\-opening this
ticket and adding it to
&nbsp;    '\#34161&#8203;: METABUG \- \(?\{\.\.\.\}\) and \(??\{\.\.\.\}\) regexp issues'

\[davem@&#8203;pigeon bleed\]$ valgrind \./perl \-Ilib /tmp/p
==4618== Memcheck\, a memory error detector
==4618== Copyright \(C\) 2002\-2009\, and GNU GPL'd\, by Julian Seward et al\.
==4618== Using Valgrind\-3\.5\.0 and LibVEX; rerun with \-h for copyright info
==4618== Command&#8203;: \./perl \-Ilib /tmp/p
==4618== 
Undefined subroutine &main&#8203;::\_noexist called at \(re\_eval 1\) line 1\.
&nbsp;    \.\.\.caught at /tmp/p line 3\.
==4618== Invalid read of size 8
==4618==    at 0x5641FA&#8203;: Perl\_pp\_const \(pp\_hot\.c&#8203;:44\)
==4618==    by 0x52154C&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2112\)
==4618==    by 0x45314E&#8203;: S\_run\_body \(perl\.c&#8203;:2310\)
==4618==    by 0x4523FC&#8203;: perl\_run \(perl\.c&#8203;:2239\)
==4618==    by 0x41F82C&#8203;: main \(perlmain\.c&#8203;:117\)
==4618==  Address 0x4c84098 is not stack'd\, malloc'd or \(recently\) free'd
==4618== 
==4618== Invalid read of size 4
==4618==    at 0x6C82EF&#8203;: Perl\_do\_print \(doio\.c&#8203;:1210\)
==4618==    by 0x56DCC1&#8203;: Perl\_pp\_print \(pp\_hot\.c&#8203;:791\)
==4618==    by 0x52154C&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2112\)
==4618==    by 0x45314E&#8203;: S\_run\_body \(perl\.c&#8203;:2310\)
==4618==    by 0x4523FC&#8203;: perl\_run \(perl\.c&#8203;:2239\)
==4618==    by 0x41F82C&#8203;: main \(perlmain\.c&#8203;:117\)
==4618==  Address 0x9c is not stack'd\, malloc'd or \(recently\) free'd


\-\- 
Please note that ash\-trays are provided for the use of smokers\,
whereas the floor is provided for the use of all patrons\.
&nbsp;    \-\- Bill Royston

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
